serving/samples/helloworld-haskell/README.md

```diff
@@ -0,0 +1,198 @@
+    version:             0.1.0.0
+    dependencies:
+    - base >= 4.7 && < 5
+    - scotty
+    - text
+
+    executables:
+      helloworld-haskell-exe:
+        main:                Main.hs
+        source-dirs:         app
+        ghc-options:
+        - -threaded
+        - -rtsopts
+        - -with-rtsopts=-N
+	```
+
+1. Create a `app` folder, then create a new file named `Main.hs` in that folder
+   and paste the following code. This code creates a basic web server which
+   listens on port 8080:
+
+    ```haskell
+	{-# LANGUAGE OverloadedStrings #-}
+
+	import           Data.Maybe
+	import           Data.Monoid        ((<>))
+	import           Data.Text.Lazy     (Text)
+	import           Data.Text.Lazy
+	import           System.Environment (lookupEnv)
+	import           Web.Scotty         (ActionM, ScottyM, scotty)
+	import           Web.Scotty.Trans
+
+	main :: IO ()
+	main = do
+      t <- fromMaybe "NOT SPECIFIED" <$> lookupEnv "TARGET"
+      scotty 8080 (route t)
+
+	route :: String -> ScottyM()
+	route t = get "/" $ hello t
+
+	hello :: String -> ActionM()
+	hello t = text $ pack ("Hello world: " ++ t)
+    ```
+
+1. In your project directory, create a file named `Dockerfile` and copy the code
+   block below into it.
+
+    ```docker
+	# Use the existing Haskell image as our base
+	FROM haskell:8.2.2 as builder
+
+	# Checkout our code onto the Docker container
+	WORKDIR /app
+	ADD . /app
+
+	# Build and test our code, then install the “helloworld-haskell-exe” executable
+	RUN stack setup
+	RUN stack build --copy-bins
+
+	# Copy the "helloworld-haskell-exe" executable to the image using docker multi stage build
+	FROM fpco/haskell-scratch:integer-gmp
+	WORKDIR /root/
+	COPY --from=builder /root/.local/bin/helloworld-haskell-exe .
+
+	# Expose a port to run our application
+	EXPOSE 8080
+
+	# Run the server command
+	CMD ["./helloworld-haskell-exe"]
+    ```
+
+1. Create a new file, `service.yaml` and copy the following service definition
+   into the file. Make sure to replace `{username}` with your Docker Hub username.
+
+    ```yaml
+    apiVersion: serving.knative.dev/v1alpha1
+    kind: Service
+    metadata:
+    name: helloworld-haskell
+    namespace: default
+    spec:
+    runLatest:
+        configuration:
+        revisionTemplate:
+            spec:
+            container:
+                image: docker.io/{username}/helloworld-haskell
+                env:
+                - name: TARGET
+                value: "Haskell Sample v1"
+    ```
+
+## Build and deploy this sample
+
+Once you have recreated the sample code files (or used the files in the sample
+folder) you're ready to build and deploy the sample app.
+
+1. Use Docker to build the sample code into a container. To build and push with
+   Docker Hub, enter these commands replacing `{username}` with your
+   Docker Hub username:
+
+    ```shell
+    # Build the container on your local machine
+    docker build -t {username}/helloworld-haskell .
```

Comment:
@abbi-gaurav 

I'm getting an error while building my container locally; here's the output:

Step 5/10 : RUN stack build --copy-bins
 ---> Running in 06e32c35baf3
Updating package index Hackage (mirrored at https://s3.amazonaws.com/hackage.fpcomplete.com/) ...
Selected mirror https://s3.amazonaws.com/hackage.fpcomplete.com/
Downloading root
Selected mirror https://s3.amazonaws.com/hackage.fpcomplete.com/
Downloading timestamp
Downloading snapshot
Downloading mirrors
Cannot update index (no local copy)
Downloading index
Updated package index downloaded
Update complete
Populating index cache ...
Populated index cache.
Error parsing targets: The specified targets matched no packages.
Perhaps you need to run 'stack init'?
The command '/bin/sh -c stack build --copy-bins' returned a non-zero code: 1

Any suggestions? I believe I've copied the sample files correctly.

Reply:
@samodell let me have a look. I will get back to you soon.

Reply:
@samodell I do not see the error when I run the command. However, let me check if adding `stack init` as suggested will help.

```bash
$ docker build -t gabbi/helloworld-haskell .
Sending build context to Docker daemon  9.219MB
Step 1/10 : FROM haskell:8.2.2 as builder
8.2.2: Pulling from library/haskell
55cbf04beb70: Pull complete
18a5c749e623: Pull complete
Digest: sha256:5bebc63cf629a42e66e87d26168dac1f82c41467e881c95220852f3e4dbbfda8
Status: Downloaded newer image for haskell:8.2.2
 ---> 0cb8be425b81
Step 2/10 : WORKDIR /app
 ---> Running in 381568b9f3aa
Removing intermediate container 381568b9f3aa
 ---> 6f96362fe9e5
Step 3/10 : ADD . /app
 ---> 5d9b40453b56
Step 4/10 : RUN stack setup
 ---> Running in ea4d4477f5f4
Downloading lts-10.7 build plan ...
Downloaded lts-10.7 build plan.
stack will use the GHC on your PATH
For more information on paths, see 'stack path' and 'stack exec env'
To use this GHC and packages outside of a project, consider using:
stack ghc, stack ghci, stack runghc, or stack exec
Removing intermediate container ea4d4477f5f4
 ---> 191b0e83fe56
Step 5/10 : RUN stack build --copy-bins
 ---> Running in bad6eb0cf548
Updating package index Hackage (mirrored at https://s3.amazonaws.com/hackage.fpcomplete.com/) ...
Selected mirror https://s3.amazonaws.com/hackage.fpcomplete.com/
Downloading root
Selected mirror https://s3.amazonaws.com/hackage.fpcomplete.com/
Downloading timestamp
Downloading snapshot
Downloading mirrors
Cannot update index (no local copy)
Downloading index
Updated package index downloaded
Update complete
Populating index cache ...
Populated index cache.
[1 of 2] Compiling Main             ( /root/.stack/setup-exe-src/setup-mPHDZzAJ.hs, /root/.stack/setup-exe-src/setup-mPHDZzAJ.o )
[2 of 2] Compiling StackSetupShim   ( /root/.stack/setup-exe-src/setup-shim-mPHDZzAJ.hs, /root/.stack/setup-exe-src/setup-shim-mPHDZzAJ.o )
Linking /root/.stack/setup-exe-cache/x86_64-linux/tmp-Cabal-simple_mPHDZzAJ_2.0.1.0_ghc-8.2.2 ...
base-compat-0.9.3: download
appar-0.1.4: download
```

Reply:
@abbi-gaurav  Oh, I re-checked my file structure and found I made the Dockerfile in the wrong location. My mistake; the build is working fine now. I'll continue walking through the guide and get back to you with edits by EOD PST.

Reply:
@samodell I confirmed that we do not need to do any changes and steps are working fine. 

Could you please try running the `docker build` command from the directory in the repository `docs/serving/samples/helloworld-haskell`?

Reply:
Was able to get through and successfully deploy.